Cleanup asserts & throws - replace with invariants - dir util/s* (part2)

sonodtt opened the pull request:
Cleanup of asserts - replaced with invariants.

Removed all unstructured throw(...) statements, replacing with INVARIANT, PRECONDITION, etc. where appropriate.

hannes-steffenhagen-diffblue commented:
Contains minor refactoring of related code.

hannes-steffenhagen-diffblue commented on src/util/simplify_expr_int.cpp:
```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
-  if(!is_bitvector_type(op0_type))
-    return true;
+  PRECONDITION(expr.id() == ID_extractbit);
+  auto &extractbit_expr = to_extractbit_expr(expr);
```
Should be `auto const&`

sonodtt replied:
Done.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -42,7 +42,10 @@ bool simplify_exprt::simplify_bswap(bswap_exprt &expr)
+      INVARIANT(
+        !bytes.empty(),
+        "bytes is not empty because we just pushed just as many elements on "
+        "top of it as we are popping now");
```
May I suggest to move this out of the loop and instead check that at least as many elements are contained in `bytes`?

sonodtt replied:
@tautschnig I'm inclined to feel that the invariant here is checking for UB in case someone makes an error in the loop condition.  
Moving the invariant out of the loop would seem to check something that is trivially true (I think...).

tautschnig replied:
Ok, not a big deal, I personally would have thought we don't need to guard against all possible future errors, but figuring out where exactly to draw this line is hard. Surely not a blocker here.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -237,7 +240,10 @@ bool simplify_exprt::simplify_mult(exprt &expr)
+    INVARIANT(
+      constant_found,
+      "c_sizeof_type is only set to a non-nil value "
+      "if a constant has been found");
```
Hmm, having two entangled variables is questionable - I know, not your design choice, but still worth looking into I'd say.

sonodtt replied:
The redesign of this section to disentangle these two variables might be considered not in the scope of the PR.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -439,6 +445,8 @@ bool simplify_exprt::simplify_mod(exprt &expr)
 
 bool simplify_exprt::simplify_plus(exprt &expr)
 {
+  PRECONDITION(expr.id() == ID_plus);
```
There's a `plus_exprt`, let's use this here.

hannes-steffenhagen-diffblue replied:
You mean `to_plus_exprt` instead of the `PRECONDITION`? I don't like that too much, the `PRECONDITION` here makes it clear we're expecting a `plus_exprt` whereas calling `to_plus_exprt` instead might be read as us just forgetting to check if it actually is...

tautschnig replied:
I was meaning to suggest that the function take a `plus_exprt &expr` as parameter. Will discuss further with `extractbit` below.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -597,14 +603,14 @@ bool simplify_exprt::simplify_plus(exprt &expr)
 
 bool simplify_exprt::simplify_minus(exprt &expr)
 {
+  PRECONDITION(expr.id() == ID_minus);
```
Use `minus_exprt`.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
-  if(!is_bitvector_type(op0_type))
-    return true;
+  PRECONDITION(expr.id() == ID_extractbit);
+  auto &extractbit_expr = to_extractbit_expr(expr);
```
Pass `extractbit_exprt` as parameter, and let `to_extractbit_expr` do all the work (at the call site) - no need for duplication.

hannes-steffenhagen-diffblue replied:
This can't take an `extractbit_exprt` because the parameter is an in/out parameter, and the result is not always going to be another `extractbit_exprt`

tautschnig replied:
We have this problem across all simplifier functions. The cleanest approach might be changing the return type from `bool` to something more elaborate so that we can return the modified expression, and would instead take a const reference as argument. Thus far, the workaround was using `swap` instead of direct assignments. I think it's important to get rid of the duplicate checks on input parameters (there should only be a check at the call site).

sonodtt replied:
@tautschnig I'm going to split this part out - and make a note for it to be considered in the context of #2939 and your remarks on #2910 (comment)

https://github.com/diffblue/cbmc/pull/2939#issuecomment-425047035

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
+  {
+    return true;
+  }
+  mp_integer index_as_int = index_converted_to_int.has_value();
```
That's ... trivially true? You might have wanted `.value()` instead of `has_value`?

hannes-steffenhagen-diffblue replied:
:flushed:

sonodtt replied:
Done.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
+    return true;
 
-  mp_integer i;
+  std::size_t src_bit_width = to_bitvector_type(src_type).get_width();
```
`const`

sonodtt replied:
Done.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
+  if(extractbit_expr.index().id() != ID_constant)
+  {
     return true;
+  }
```
I believe this check is unnecessary as the subsequent `numeric_cast<mp_integer>` would catch this case anyway.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
 
-  bool bit=(id2string(value)[width-integer2size_t(i)-1]=='1');
+  bool bit =
+    (id2string(src_value)
```
Do `id2string` earlier on as doing `src_value.size()` will also require building the string.

sonodtt replied:
Done.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 99b2b26b5ba825b3c32421ff18bb2ecdd0d71380).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84018230

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: d12acb32bc5c5a89e2314a093fc22ac93de6ff03).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84135814
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -41,6 +41,10 @@ bool simplify_exprt::simplify_bswap(bswap_exprt &expr)
+
+    //May I suggest to move this out of the loop and instead check that at least as many elements are contained in bytes?
+      // this invariant protects against someone modifying the loop
+        // expression - a bit pointless for it to be outside the loop.
```
I don't think this belongs in here?

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -43,8 +43,8 @@ bool simplify_exprt::simplify_bswap(bswap_exprt &expr)
-      // this invariant protects against someone modifying the loop
-        // expression - a bit pointless for it to be outside the loop.
+    // this invariant protects against someone modifying the loop
+    // expression - a bit pointless for it to be outside the loop.
```
I'd get rid of this.

sonodtt replied:
Apologies @tautschnig - this was an internal note for a colleague for a second opinion. 
I was used to a workflow where I'd push to get CI going, review, then re-add reviewers to flag ready for re-review. 
You often re-review quite quickly - in future will push after local checks.

tautschnig replied:
Oh don't worry, I was just naively assuming this was ready for re-review.

Commit:
Cleanup asserts & throws - replace with invariants

* C-style asserts are deprecated in favor of CBMCs INVARIANT macros
* Some minor refactoring that was relevant to the cleanup.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d3a148de87dc48fc90f603ea01bd8baa6903af29).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86186513

hannes-steffenhagen-diffblue committed:
Refactor simplify_minus

* Use to_minus_expr instead of asserting the expression id
* Rename the local variables for easier readability

hannes-steffenhagen-diffblue committed:
Use to_plus_expr instead of assert in simplify_plus

to_plus_expr should check if an expression is a plus_exprt, this avoids
duplicating checks across the codebase

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: bd2558b5277fba4b88de4b4211421c1c57b3aad6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86464235

hannes-steffenhagen-diffblue commented:
@tautschnig Can you rereview? There is a clang-format failure that I can't fix at the moment
 
~~(I'm not at my usual workstation and the version of clang-format I have here seems to disagree with what's running on travis), so I wouldn't merge it yet, but I could just easily fix that tomorrow~~

NVM the above part, it's been rightfully pointed out to me that not having a one line change be done automatically for you is a stupid reason to delay it.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d5837ad56665830e5ea87dc5362b60499254880f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86491547

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -604,43 +604,42 @@ bool simplify_exprt::simplify_plus(exprt &expr)
 
 bool simplify_exprt::simplify_minus(exprt &expr)
 {
+  auto const &minus_expr = to_minus_expr(expr);
```
The lines after this one still use `expr`, not `minus_expr`

hannes-steffenhagen-diffblue replied:
I've made the functions use them where it seemed reasonable.

tautschnig commented on src/util/simplify_expr_int.cpp:
```diff
@@ -446,6 +446,8 @@ bool simplify_exprt::simplify_mod(exprt &expr)
 bool simplify_exprt::simplify_plus(exprt &expr)
 {
+  // only used to check structural invariants
+  (void)to_plus_expr(expr);
```
Maybe still use this, in particular the const version thereof, in the next lines? Doing `.type()` on a `const` expression is cheaper, thus beneficial.

tautschnig reviewed (approved):
Just a few minor suggestions.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 07f7d1e99011a668ac5b8fb635902803bdc2295a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86591624
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 69e5c2d7f31abe0e286b105f3b281624dc890cd2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86592975
